de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java

```diff
@@ -532,44 +550,104 @@ public void addResourcesToSession(
             return;
         }
 
-        List<IProject> projectsToShare = new ArrayList<IProject>();
+        /** add new shared resources to a combining map for the next negotiation **/
+        synchronized (this) {
+            for (Entry<IProject, List<IResource>> mapEntry : projectResourcesMapping
+                .entrySet()) {
+                final IProject project = mapEntry.getKey();
+                final List<IResource> resourcesList = mapEntry.getValue();
+
+                /** case: first entry for project **/
+                if (!nextSharedResources.containsKey(project)) {
+                    if (resourcesList == null || resourcesList.isEmpty())
+                        nextSharedResources.put(project, null);
+                    else {
+                        nextSharedResources.put(project,
+                            new ArrayList<IResource>(resourcesList));
+                    }
+                } else {
+                    List<IResource> list = nextSharedResources.get(project);
+
+                    /** case: project is already a completely shared **/
+                    if (list == null || list.isEmpty())
+                        continue;
+                    /** case: update partial project to completely shared **/
+                    if (resourcesList == null || resourcesList.isEmpty())
+                        nextSharedResources.put(project, null);
+                    else
+                        /** case: just add new resources to a queued project **/
+                        list.addAll(resourcesList);
+                }
+            }
+
+            /** leave if a thread is already waiting to do the work **/
+            if (nextSharedResourcesWorker) {
+                return;
+            }
+            /** otherwise become the working thread **/
+            nextSharedResourcesWorker = true;
+        }
+
+        /**
+         * leave the synchronized block to allow new resource adding and wait
+         * till all running negotiations are done
+         **/
+        synchronized (nextSharedResourcesWorkerNotify) {
+            while (currentProjectNegotiations.negotiationsInProgress()) {
+                try {
+                    nextSharedResourcesWorkerNotify.wait();
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
```

Comment:
return ?

Reply:
would break the logic, only case escaping this would be a ending session or not?

Reply:
final Object l = new Object();
		int x = 0;
		
		Thread.currentThread().interrupt();
		
		synchronized(l) {
			while(x < 1000) {
				try {
					l.wait();
				} catch (InterruptedException e) {
					System.err.println("interrupted");
					Thread.currentThread().interrupt();
					x++;
				}
			}
		}
translate into busy spinning

If you want to ignore the interruption then catch the exception, remember that you got interrupted and when your logic is done call Thread#interrupt();

However getting interrupted is a signal to stop the thread gracefully, and so it should terminate as early as it can.